docs(cn): translate reference/react/useRef into Chinese (#1132)
reactjs/zh-hans.react.dev: src/content/reference/react/useRef.md

  // ...
```

[请看下面的更多例子](#usage)

#### 参数 {/*parameters*/}

* `initialValue`：ref 对象的 `current` 属性的初始值。可以是任意类型的值。这个参数会首次渲染后被忽略。

#### 返回值 {/*returns*/}

`useRef` 返回一个只有一个属性的对象:

* `current`：最初，它被设置为你传递的 `initialValue`。之后你可以把它设置为其他值。如果你把 ref 对象作为一个 JSX 节点的 `ref` 属性传递给 React，React 将为它设置`current`属性。

在接下来渲染中，`useRef` 将返回同一个对象。

#### 注意事项 {/*caveats*/}

* 你可以修改 `ref.current` 属性。与 state 不同，它是可变的。然而，如果它持有一个用于渲染的对象（例如，你的 state 的一部分），那么你就不应该修改这个对象。
* 当你改变 `ref.current` 属性时，React 不会重新渲染你的组件。React 不知道你何时改变它，因为 ref 是一个普通的 JavaScript 对象。
* 除了[初始化](#avoiding-recreating-the-ref-contents)外不要在渲染期间写入 _或者读取_ `ref.current`。这会使你的组件的行为不可预测。

> * 在严格模式下，React 将会 **调用两次你的组件方法**，这是为了[帮助你发现意外的问题](#my-initializer-or-updater-function-runs-twice)。这只是开发模式下的行为，不影响生产模式。每个 ref 对象将会创建两次，但是其中一个版本将被丢弃。如果你的组件函数是纯的（应该如此），这不会影响其行为。

Reviewer: ```suggestion
* 在严格模式下，React 将会 **调用两次你的组件方法**，这是为了 [帮助你发现意外的问题](#my-initializer-or-updater-function-runs-twice)。这只是开发模式下的行为，不影响生产模式。每个 ref 对象将会创建两次，但是其中一个版本将被丢弃。如果你的组件函数是纯的（应该如此），这不会影响其行为。
```

}
```

通过使用 ref，你可以确保：

- 你可以在重新渲染之间 **存储信息**（不像是普通对象，每次渲染都会重置）。
- 改变它 **不会触发重新渲染**（不像是 state 变量，会触发重新渲染）。
- 对于你的组件的每个副本来说，**这些信息都是本地的**（不像是外面的变量，是共享的）。

> 改变 ref 不会触发重新渲染，所以 refs 不适合存储用于你想要显示在屏幕上的信息。使用 state 代替。阅读更多[在 `useRef` 和 `useState` 之间选择。](/learn/referencing-values-with-refs#differences-between-refs-and-state)

Reviewer 2: ```suggestion
改变 ref 不会触发重新渲染，所以 ref 不适合存储用于你想要显示在屏幕上的信息。如有需要，使用 state 代替。阅读更多 [在 `useRef` 和 `useState` 之间选择](/learn/referencing-values-with-refs#differences-between-refs-and-state)。
```

<Pitfall>

**不要在渲染期间写入 _或者读取_ `ref.current`。**

> React 期望你的组件的主体[表现得像一个纯函数](/learn/keeping-components-pure)：

Reviewer: ```suggestion
React 期望你的组件的主体 [表现得像一个纯函数](/learn/keeping-components-pure)：
```

- 如果输入([props](/learn/passing-props-to-a-component)、[state](/learn/state-a-components-memory) 和 [context](/learn/passing-data-deeply-with-context))都是一样的，那么就应该返回一样的 JSX。
- 以不同的顺序或用不同的参数调用它，不应该影响其他调用的结果。

在 **渲染期间** 读取或写入 ref 会破坏这些预期。

```js {3-4,6-7}
function MyComponent() {
  // ...
  // 🚩 不要在渲染期间写入 ref
  myRef.current = 123;
  // ...
  // 🚩 不要在渲染期间读取 ref
  return <h1>{myOtherRef.current}</h1>;
}
```

> 你可以在 **事件处理函数或者 effects** 中读取和写入 refs。

Reviewer 2: ```suggestion
你可以在 **事件处理函数或者 effects** 中读取和写入 ref。
```

```js {4-5,9-10}
function MyComponent() {
  // ...
  useEffect(() => {

>     // ✅ 你可以在 effects 中读取和写入 refs

Reviewer 2: ```suggestion
    // ✅ 你可以在 effects 中读取和写入 ref
```

    myRef.current = 123;
  });
  // ...
  function handleClick() {

>     // ✅ 你可以在事件处理函数中读取和写入 refs

Reviewer 2: ```suggestion
    // ✅ 你可以在事件处理函数中读取和写入 ref
```

    doSomething(myOtherRef.current);
  }
  // ...
}
```

如果 *不得不* 在渲染期间读取[或者写入](/reference/react/useState#storing-information-from-previous-renders)，[使用 state](/reference/react/useState) 代替。

> 当你打破这些规则时，你的组件可能仍然可以工作，但我们添加到 React 的大多数新功能将依赖于这些预期行为。阅读更多关于[保持你的组件纯粹。](/learn/keeping-components-pure#where-you-can-cause-side-effects)

Reviewer: ```suggestion
当你打破这些规则时，你的组件可能仍然可以工作，但我们添加到 React 的大多数新功能将依赖于这些预期行为。阅读 [保持你的组件纯粹](/learn/keeping-components-pure#where-you-can-cause-side-effects) 以了解更多信息。
```。

</Pitfall>

---

### 通过 ref 操作 DOM {/*manipulating-the-dom-with-a-ref*/}

> 使用 ref 操作 [DOM](https://developer.mozilla.org/zh-CN/docs/Web/API/HTML_DOM_API) 是非常常见的。React 对此有内置支持。

Reviewer: ```suggestion
使用 ref 操作 [DOM](https://developer.mozilla.org/zh-CN/docs/Web/API/HTML_DOM_API) 是非常常见的。React 内置了对它的支持。
```

<Pitfall>

**不要在渲染期间写入 _或者读取_ `ref.current`。**

React 期望你的组件的主体 [表现得像一个纯函数](/learn/keeping-components-pure)：

- 如果输入([props](/learn/passing-props-to-a-component)、[state](/learn/state-a-components-memory) 和 [context](/learn/passing-data-deeply-with-context))都是一样的，那么就应该返回一样的 JSX。
- 以不同的顺序或用不同的参数调用它，不应该影响其他调用的结果。

在 **渲染期间** 读取或写入 ref 会破坏这些预期。

```js {3-4,6-7}
function MyComponent() {
  // ...
  // 🚩 不要在渲染期间写入 ref
  myRef.current = 123;
  // ...
  // 🚩 不要在渲染期间读取 ref
  return <h1>{myOtherRef.current}</h1>;
}
```

你可以在 **事件处理函数或者 effects** 中读取和写入 ref。

```js {4-5,9-10}
function MyComponent() {
  // ...
  useEffect(() => {
    // ✅ 你可以在 effects 中读取和写入 ref
    myRef.current = 123;
  });
  // ...
  function handleClick() {
    // ✅ 你可以在事件处理函数中读取和写入 ref
    doSomething(myOtherRef.current);
  }
  // ...
}
```

如果 *不得不* 在渲染期间读取[或者写入](/reference/react/useState#storing-information-from-previous-renders)，[使用 state](/reference/react/useState) 代替。

> 当你打破这些规则时，你的组件可能仍然可以工作，但我们添加到 React 的大多数新功能将依赖于这些预期行为。阅读 [保持你的组件纯粹](/learn/keeping-components-pure#where-you-can-cause-side-effects) 以了解更多信息。

Reviewer: ```suggestion
当你打破这些规则时，你的组件可能仍然可以工作，但我们添加到 React 的大多数新功能将依赖于这些预期行为。阅读 [保持你的组件纯粹](/learn/keeping-components-pure#where-you-can-cause-side-effects) 以了解更多信息。

```

> 

Reviewer: ```suggestion
```

<Pitfall>

**不要在渲染期间写入 _或者读取_ `ref.current`。**

React 期望你的组件的主体 [表现得像一个纯函数](/learn/keeping-components-pure)：

- 如果输入([props](/learn/passing-props-to-a-component)、[state](/learn/state-a-components-memory) 和 [context](/learn/passing-data-deeply-with-context))都是一样的，那么就应该返回一样的 JSX。
- 以不同的顺序或用不同的参数调用它，不应该影响其他调用的结果。

在 **渲染期间** 读取或写入 ref 会破坏这些预期。

```js {3-4,6-7}
function MyComponent() {
  // ...
  // 🚩 不要在渲染期间写入 ref
  myRef.current = 123;
  // ...
  // 🚩 不要在渲染期间读取 ref
  return <h1>{myOtherRef.current}</h1>;
}
```

你可以在 **事件处理函数或者 effects** 中读取和写入 ref。

```js {4-5,9-10}
function MyComponent() {
  // ...
  useEffect(() => {
    // ✅ 你可以在 effects 中读取和写入 ref
    myRef.current = 123;
  });
  // ...
  function handleClick() {
    // ✅ 你可以在事件处理函数中读取和写入 ref
    doSomething(myOtherRef.current);
  }
  // ...
}
```

如果 *不得不* 在渲染期间读取[或者写入](/reference/react/useState#storing-information-from-previous-renders)，[使用 state](/reference/react/useState) 代替。

当你打破这些规则时，你的组件可能仍然可以工作，但我们添加到 React 的大多数新功能将依赖于这些预期行为。阅读 [保持你的组件纯粹](/learn/keeping-components-pure#where-you-can-cause-side-effects) 以了解更多信息。

> ```。

Reviewer: ```suggestion
```
```

  // ...
```

然后将你的 ref 对象作为 `ref` 属性传递给你想要操作的 DOM 节点的 JSX：

```js [[1, 2, "inputRef"]]
  // ...
  return <input ref={inputRef} />;
```

在 React 创建了 DOM 节点并将其置于屏幕后，React 将会把你的 ref 对象的 <CodeStep step={2}>`current` 属性</CodeStep>设置为那个 DOM 节点。现在你可以访问 `<input>` 的 DOM 节点，并且可以调用类似于 [`focus()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus) 的方法：

```js [[2, 2, "inputRef.current"]]
  function handleClick() {
    inputRef.current.focus();
  }
```

当节点从屏幕上移除时，React 将把 `current` 属性设回 `null`。

> 阅读更多关于[用 ref 操纵DOM](/learn/manipulating-the-dom-with-refs)

Reviewer 2: ```suggestion
阅读更多关于 [用 ref 操纵DOM](/learn/manipulating-the-dom-with-refs)
```

> 阅读更多关于 [用 ref 操纵DOM](/learn/manipulating-the-dom-with-refs)

Reviewer: ```suggestion
阅读 [用 ref 操纵 DOM](/learn/manipulating-the-dom-with-refs) 以了解更多信息。
```

<Solution />

#### 向你的组件暴露 ref {/*exposing-a-ref-to-your-own-component*/}

> 有时，你可能想让父级组件在你的组件中操纵 DOM。例如，也许你正在写一个 `MyInput` 组件，但你希望父组件能够聚焦 input（父组件无法访问）。你可以使用一个组合，通过 `useRef` 持有 input 并且通过 [`forwardRef`](/reference/react/forwardRef) 来将其暴露给父组件。在这里阅读一个[详细演练](/learn/manipulating-the-dom-with-refs#accessing-another-components-dom-nodes)。

Reviewer 2: ```suggestion
有时，你可能想让父级组件在你的组件中操纵 DOM。例如，也许你正在写一个 `MyInput` 组件，但你希望父组件能够聚焦 input（父组件无法访问）。你可以使用一个组合，通过 `useRef` 持有 input 并且通过 [`forwardRef`](/reference/react/forwardRef) 来将其暴露给父组件。在这里阅读一个 [详细演练](/learn/manipulating-the-dom-with-refs#accessing-another-components-dom-nodes)。
```

---

### 避免重复创建 ref 的内容 {/*avoiding-recreating-the-ref-contents*/}

> React 保存首次的 ref 初始值，并在后续的渲染中忽略他。

Reviewer 2: ```suggestion
React 保存首次的 ref 初始值，并在后续的渲染中忽略它。
```

  // ...
```

通常情况下，在渲染过程中写入或读取 `ref.current` 是不允许的。然而，在这种情况下是可以的，因为结果总是一样的，而且条件只在初始化时执行，所以是完全可以预测的。

<DeepDive>

> #### 如何避免在初始化 useRef 之后进行 null 的类型检查 {/*how-to-avoid-null-checks-when-initializing-use-ref-later*/}

Reviewer 2: ```suggestion
#### 如何避免在初始化 `useRef` 之后进行 `null` 的类型检查 {/*how-to-avoid-null-checks-when-initializing-use-ref-later*/}
```

  // ...
```

> 在这里，`playerRef` 本身是可以为空的。然而，你应该能够说服你的类型检查器，不存在 `getPlayer()` 返回 `null` 的情况。然后在你的事件处理程序中使用 `getPlayer()`。

Reviewer: ```suggestion
在这里，`playerRef` 本身是可以为空的。然而，你应该能够使你的类型检查器确信，不存在 `getPlayer()` 返回 `null` 的情况。然后在你的事件处理程序中使用 `getPlayer()`。
```

</DeepDive>

---

## 疑难解答 {/*troubleshooting*/}

### 我无法获取自定义组件的 ref {/*i-cant-get-a-ref-to-a-custom-component*/}

如果你尝试向这样传递 `ref` 到你自己的组件：

```js
const inputRef = useRef(null);

return <MyInput ref={inputRef} />;
```

你可能会在控制台中得到这样的错误：

<ConsoleBlock level="error">

Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?

</ConsoleBlock>

> 默认情况下，你自己的组件不会暴露它们内部的 DOM 节点的 refs。

Reviewer 2: ```suggestion
默认情况下，你自己的组件不会暴露它们内部的 DOM 节点的 ref。
```

}
```

> 然后像这样将其包装在 [`forwardRef`](/reference/react/forwardRef)：

Reviewer: ```suggestion
然后像这样将其包装在 [`forwardRef`](/reference/react/forwardRef) 里：
```

export default MyInput;
```

然后，父级组件可以得到它的 ref。

> 阅读更多关于[访问另一个组件的 DOM 节点](/learn/manipulating-the-dom-with-refs#accessing-another-components-dom-nodes)。

Reviewer 2: ```suggestion
阅读更多关于 [访问另一个组件的 DOM 节点](/learn/manipulating-the-dom-with-refs#accessing-another-components-dom-nodes)。
```

> 阅读更多关于 [访问另一个组件的 DOM 节点](/learn/manipulating-the-dom-with-refs#accessing-another-components-dom-nodes)。

Reviewer: ```suggestion
阅读 [访问另一个组件的 DOM 节点](/learn/manipulating-the-dom-with-refs#accessing-another-components-dom-nodes) 了解更多信息。
```
